File: cpp/0501二叉搜索树的众数.cpp
#include "mybitree.h"
#include <stack>
#include <vector>
#include <algorithm>

vector<int> findMode(TreeNode *root)
{
    stack<TreeNode *> st;
    st.push(root);
    int temp = 0xffffffff;
    vector<int> res;
    int count = 0;
    int max_count = 0;
    vector<TreeNode *> memo;
    while (!st.empty())
    {
        TreeNode *crt = st.top();

        if (crt->left && find(memo.begin(), memo.end(), crt->left) == memo.end())
        {
            st.push(crt->left);
            continue;
        }
        else
        {
            st.pop();
            if (temp != crt->val)
                count = 0;
            temp = crt->val;
            ++count;

            if (count == max_count)
                res.push_back(temp);

            if (count > max_count)
            {
                max_count = count;
                res.clear();
                res.push_back(temp);
            }

            memo.push_back(crt);
        }
        if (crt->right)
            st.push(crt->right);
    }
    return res;
}

File: python/main/0501二叉搜索树的众数.py
from binarytree import TreeNode,level_create


def findMode(root: TreeNode) -> list:

    counts=[0,0]
    res=[]
    pre=None

    def _traversal(root:TreeNode):
        nonlocal pre
        if(root):
            _traversal(root.left)
            
            if(pre==None or pre.val==root.val):
                counts[0]+=1
            else:
                counts[0]=1

            pre=root
            if(counts[0]==counts[1]):
                res.append(root.val)
            
            if(counts[0]>counts[1]):
                counts[1]=counts[0]
                res.clear()
                res.append(root.val)
            
            _traversal(root.right)

    _traversal(root)
    return res

print(findMode(level_create([0])))